lib/smack/SmackInstGenerator.cpp

```diff
@@ -882,4 +832,167 @@ void SmackInstGenerator::visitMemSetInst(llvm::MemSetInst& msi) {
+    if (SmackOptions::FloatEnabled)
+      emit(Stmt::assign(rep->expr(ci),
+        Expr::fn(indexedName("$fma",
+          {rep->type(ci->getFunctionType()->getReturnType())}),
+        rep->expr(ci->getArgOperand(0)),
+        rep->expr(ci->getArgOperand(1)),
+        rep->expr(ci->getArgOperand(2)))));
+    else
+      generateUnModeledCall(ci);
+  };
+
+  static const auto bswap = [this](CallInst* ci){
+    if (SmackOptions::BitPrecise) {
+      auto width = ci->getArgOperand(0)->getType()->getIntegerBitWidth();
+      auto var = rep->expr(ci->getArgOperand(0));
+
+      // Swap the bytes to the right and left of the middle
+      const Expr* body = Expr::bvConcat(Expr::bvExtract(var,width/2,width/2-8),
+                                        Expr::bvExtract(var,width/2+8,width/2));
+
+      // Swap the bytes to the right and the left of the already swapped portion.
+      for (unsigned i = 8; i < width/2; i += 8) {
+        body = Expr::bvConcat(Expr::bvConcat(Expr::bvExtract(var,width/2-i,width/2-i-8),
+                                             body),
+                              Expr::bvExtract(var, width/2+i+8, width/2+i));
+      }
+      emit(Stmt::assign(rep->expr(ci), body));
+    }
+    else
+      generateUnModeledCall(ci);
+  };
+
+  static const auto assignBinFPFuncApp = [this] (CallInst* ci, std::string fnBase) {
+    // translation: $res := $<func>.bv*($arg1, $arg2);
+    if (SmackOptions::FloatEnabled)
+      emit(Stmt::assign(
+            rep->expr(ci),
+            Expr::fn(indexedName(fnBase,
+                {rep->type(ci->getFunctionType()->getReturnType())}),
+              {rep->expr(ci->getArgOperand(0)), rep->expr(ci->getArgOperand(1))})));
+    else
+      generateUnModeledCall(ci);
+  };
+
+  static const auto assignRoundFPFuncApp = [this] (CallInst* ci, const Expr* rMode) {
+    // translation: $res := $round.bv*(rmode, $arg1);
+    if (SmackOptions::FloatEnabled)
+      emit(Stmt::assign(
+            rep->expr(ci),
+            Expr::fn(indexedName("$round",
+                {rep->type(ci->getFunctionType()->getReturnType())}),
+              {rMode, rep->expr(ci->getArgOperand(0))})));
+    else
+      generateUnModeledCall(ci);
+  };
+
+  static const auto identity = [this] (CallInst* ci) {
+    // translation: $res := $arg1
+    Value* val = ci->getArgOperand(0);
+    emit(Stmt::assign(rep->expr(ci), rep->expr(val)));
+  };
+
+  // TODO: these functions is consistent with the implementations in math.c,
+  // meaning we can use __builtin_* to implement math.c which is mostly
+  // modeled using __SMACK_code.
+
+  static const std::map<llvm::Intrinsic::ID, std::function<void(CallInst*)>> stmtMap {
+    {llvm::Intrinsic::convert_from_fp16, f16UpCast},
+    {llvm::Intrinsic::convert_to_fp16, f16DownCast},
+    {llvm::Intrinsic::bswap, bswap},
+    {llvm::Intrinsic::expect, identity},
+    {llvm::Intrinsic::fabs,
+      [] (CallInst* ci){ assignUnFPFuncApp(ci,"$abs"); }},
```

Comment:
Seems like you could simplify the following cases quite a bit by currying. For instance, rather than defining the `assignUnFPFuncApp` closure above, define the following function:
```C++
std::function<void(CallInst*)> SmackInstGenerator::assignUnFPFuncApp(std::string fnBase) {
  return [this, fnBase] (CallInst* ci) {
    // translation: $res := $<func>.bv*($arg1);
    if (SmackOptions::FloatEnabled)
      emit(Stmt::assign(
            rep->expr(ci),
            Expr::fn(indexedName(fnBase,
                {rep->type(ci->getArgOperand(0)->getType())}),
              rep->expr(ci->getArgOperand(0)))));
    else
      generateUnModeledCall(ci);
  };
}
```
Then you could replace the current line with simply:
```C++
assignUnFPFuncApp("$abs")},
```

Reply:
Good suggestion. Is there a particular reason that this function is made as a member method instead of a local variable?

Reply:
That’s probably just stylistic preferences; either way resolves the issue.